src/main/java/com/glencoesoftware/omero/ms/image/region/ImageRegionVerticle.java

```diff
@@ -138,45 +292,311 @@ private void renderImageRegion(Message<String> message) {
+     * @param imageRegionCtx request context
+     * @param request OMERO request based on the current context
+     * @return Future which will be completed with the image region byte array.
+     */
+    private Future<byte[]> getImageRegion(
+            ImageRegionCtx imageRegionCtx, Supplier<OmeroRequest> request) {
+        Summary.Timer timer = getImageRegionSummary.startTimer();
+        Future<byte[]> future = Future.future();
+
+        try {
             if (families == null) {
-                request.execute(this::updateFamilies);
+                request.get().execute(this::updateFamilies);
             }
             if (renderingModels == null) {
-                request.execute(this::updateRenderingModels);
+                request.get().execute(this::updateRenderingModels);
             }
+        } catch (Exception e) {
+            future.fail(e);
+            return future;
+        }
+
+        Future<byte[]> step1 = getCachedImageRegion(imageRegionCtx, request);
+
+        step1.setHandler(result1 -> {
+            if (result1.succeeded()) {
+                // If the region is in the cache complete and return
+                byte[] imageRegion = result1.result();
+                if (imageRegion != null) {
+                    future.complete(imageRegion);
+                } else {
+                    // The region is not in the cache, we have to create it
+                    Future<byte[]> step2 =
+                            renderImageRegion(imageRegionCtx, request);
 
-            byte[] imageRegion = request.execute(
-                    new ImageRegionRequestHandler(
-                            imageRegionCtx, context, families,
-                            renderingModels, lutProvider)::renderImageRegion);
-            if (imageRegion == null) {
-                message.fail(
-                        404, "Cannot find Image:" + imageRegionCtx.imageId);
+                    step2.setHandler(result2 -> {
+                        if (result2.succeeded()) {
+                            future.complete(result2.result());
+                        } else {
+                            future.fail(result2.cause());
+                        }
+                    });
+                }
             } else {
-                message.reply(imageRegion);
+                future.fail(result1.cause());
             }
-        } catch (PermissionDeniedException
-                | CannotCreateSessionException e) {
-            String v = "Permission denied";
-            log.debug(v);
-            message.fail(403, v);
-        } catch (IllegalArgumentException e) {
-            log.debug(
-                "Illegal argument received while retrieving image region", e);
-            message.fail(400, e.getMessage());
-        } catch (Exception e) {
-            String v = "Exception while retrieving image region";
-            log.error(v, e);
-            message.fail(500, v);
+        });
+
+        timer.observeDuration();
```

Comment:
With all the async code above, observing the duration here won't tell us anything, it will end up being 0ms.